### PR TITLE
Better power net

### PR DESCRIPTION
**Pull request by botn365**
change the power net to a graph based search to make it more tps frendly

### CODE REVIEW

**Review by Glease — CHANGES_REQUESTED** [review]

**Comment by Glease** [comment]
After reading through `PowerNodes`, I still don't quite understand the role of node value here. Can you please briefly explain the purpose of this thing?

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java [hunk 1]
```diff
@@ -169,118 +173,30 @@ public long transferElectricity(byte aSide, long aVoltage, long aAmperage, Array
...
-                    rUsedAmperes++;
-                    mRestRF = rfOut - RFtrans;
-                } else {
-                    int RFtrans = rfReceiver.receiveEnergy(tDirection, (int) mRestRF, false);
-                    mRestRF = mRestRF - RFtrans;
+        PowerNode tNode =(PowerNode) tBase.getNode();
+        if (tNode != null) {
+            int tPlace = 0;
+            Node[] tToPower = new Node[tNode.mConsumers.size()];
+            if (tNode.mHadVoltage) {
```
[the remark this answers is not in the document]
Comment (botn365) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java:184]: yes it is intentional its there to aplly voltage to the network when it has been reset
i dont know of a better solution then a array

File: src/main/java/gregtech/api/graphs/paths/PowerNodePath.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                    BaseMetaPipeEntity tBaseCable = (BaseMetaPipeEntity) tCable.getBaseMetaTileEntity();
+                    tBaseCable.setToFire();
+                }
+            }
+        }
+    }
+
+    public void addAmps(int aAmps) {
+        this.mAmps += aAmps;
+        if (false && this.mAmps > mMaxAmps * 40) {
```
[the remark this answers is not in the document]
Comment (botn365) [reply src/main/java/gregtech/api/graphs/paths/PowerNodePath.java:51]: debug code i forgot to remove

File: src/main/java/gregtech/api/metatileentity/BaseMetaTileEntity.java [hunk 3]
```diff
@@ -1098,6 +1111,29 @@ public boolean outputsEnergyTo(byte aSide, boolean waitForActive) {
...
+                    IGregTechTileEntity TE = getIGregTechTileEntityAtSide(i);
+                    if (TE instanceof BaseMetaPipeEntity) {
+                        Node node = ((BaseMetaPipeEntity) TE).getNode();
+                        if (node == null) {
+                            new GenerateNodeMapPower((BaseMetaPipeEntity) TE);
+                        } else if (node.mCreationTime != time) {
+                            GenerateNodeMap.clearNodeMap(node,-1);
+                            new GenerateNodeMapPower((BaseMetaPipeEntity) TE);
+                        } else {
+//                            GenerateNodeMap.clearNodeMap(node,-1);
```
[the remark this answers is not in the document]
Comment (botn365) [reply src/main/java/gregtech/api/metatileentity/BaseMetaTileEntity.java:1128]: debug code i forgot to remove

**Comment by botn365** [comment]
its used to find the path to the node it needs to power

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Glease) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java:184]: What if you just pass the mConsumers directly to the PowerNodes? Except during first tick, it's really just copying the same list several dozen times per tick, depending on how many power producers you have on this same network.

[the remark this answers is not in the document]
Comment (botn365) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java:184]: nope its only a list of consumers that need power 
consumers that dont need power dont get put in the list

[the remark this answers is not in the document]
Comment (Glease) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java:184]: Then you are missing a if guard below, in the else branch.

[the remark this answers is not in the document]
Comment (botn365) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java:184]: if (consumer.needsEnergy())
    tToPower[tPlace++] = consumer;

this works even without the {} if it is only 1 line

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java [hunk 4]
```diff
@@ -169,118 +173,30 @@ public long transferElectricity(byte aSide, long aVoltage, long aAmperage, Array
...
+                for (ConsumerNode consumer : tNode.mConsumers) {
+                    if (consumer.needsEnergy())
+                        tToPower[tPlace++] = consumer;
+                }
+            } else {
+                tNode.mHadVoltage = true;
+                for (ConsumerNode consumer : tNode.mConsumers) {
+                    tToPower[tPlace++] = consumer;
```
Comment (Glease) [inline src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java:192]: Look here. I'm talking about this else branch.

Comment (botn365) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java:192]: that only gets called once after a new node network was genarated

Comment (botn365) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java:192]: and toPower is a array while consumers is a arraylist

Comment (Glease) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaPipeEntity_Cable.java:192]: Oh, yeah stupid me. Then we are clear.

**Review by Glease — APPROVED** [review]
No obvious bugs per se. Let's test this on zeta asap. We also have structurelib integration going on and I don't want to test two monumental changes all at the same time.

**Comment by botn365** [comment]
would be nice to have some before and after tps data on the enet

**Comment by boubou19** [comment]
with this on zeta, we roughly divided by 4 the load caused by the cables with the same bases loaded in memory (from 4.48% to 1.39% of cpu load). So far no bugs have been seen, but let's wait a couple of few days before claiming it safe.

**Pull request merged**